tests/tests_app/core/test_lightning_flow.py

```diff
@@ -634,3 +637,28 @@ def run(self):
             assert len(self._calls["scheduling"]) == 8
 
     Flow().run()
+
+
+class CounterDB(LightningSpec, table=True):
+    id: Optional[int] = Field(default=None, primary_key=True)
+    counter: int = 0
+
+
+class FlowDB(LightningFlow):
+    def __init__(self):
+        super().__init__()
+        self.counter = CounterDB()
+        self.counter.reload()
+
+    def run(self):
+        if self.counter.counter >= 1000:
+            self._exit()
+
+        print(self.counter.counter)
+        self.counter.counter += 1
+
+
+def test_lightning_flow_sql_model():
+
+    app = LightningApp(FlowDB(), database="sqlite")
```

Comment:
passing the `database` argument to `LightningApp` is very disconnected from initializing the DB object inside the Flow. I don't have a solution to recommend for that but we might need to reconsider

Reply:
I think we can avoid passing it to LightningApp. the `self.db` property could be a subclass of a baseDB and contain the needed info for the db.

We can do something like this ?

```python


class LightningBaseDB:
     .....

class LightningDefaultSqliteDB(LightningBaseDB):
    # stored as a file in the drive
    .....

class LightningPostgresComponent(LightningBaseDB):
   .....


class Table1(Present):
    id: Optional[int] = Field(default=None, primary_key=True)
    count: int = 0

class Table2(Present): 
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str

class MySimpleDB(LightningBaseDB):
    # schema
    table1: Table1
    table2: Table2


class MyApp(LightningFlow):
    def __init__(self):
        super().__init__()

        self.db = MySimpleDB()
        self.db.reload()

    def run(self):
        if self.db.tabel1.counter >= 1000:
            self._exit()

        print(self.counter.counter)
        self.db.tabel1.counter += 1


def test_lightning_flow_sql_model():

    app = LightningApp(FlowDB())

```

Reply:
Hey @hhsecond @manskx.

I will think about the proposition above. IMO, it isn't exactly what users need.

They need a simple way to persist a component state to a database in an atomic way and be able to restart from that state. In this PoC PR, the state is persisted with the component name and can be reloaded.

I like being able to access the tables in a structured way, but the counter is raw within the table, not the table entirely.

Reply:
@lantiga @awaelchli What are your thoughts? I would be curious to see someone else PoC on this subject. I think it is one of the most important API to ace from the framework.

Reply:
I think we still don’t have clarity on the purpose.

I see two disjoint needs here, we need to tackle them as such, although at some point we will see the two needs addressed by the same solution, but it’s too early to say:
- providing a way for users to persist information (not state), for instance jobs they historically launched. When I say “not state” I mean that state might still hold a reference to a particular snapshot in time of such db, as a whole.
- providing a durable recording of state and its history, potentially with transactional semantics, a-la temporal.io

I strongly recommend we take this PR as a design brainstorm. We are not ready to commit to a design or the other.

I’ll make more specific comments asap.